hello-world/2023-02-20_testable_clock.md

```diff
@@ -0,0 +1,104 @@
+﻿# A testable clock
+Every developer that ever build an application that has any time dependency
+knows that testing such dependencies is not straight forward. The most pressing
+issue is arguably that most programming languages (including C#, PHP, Java, SQL)
+do not allow you to change the behaviour of the (system) clock/date time provider.
+So how to overcome this problem?
+
+## Stubable Date Time provider
+With the addoption of depdency injection/inverse of control, most programmers
+have developed a habbit of creating services for everything. DI frameworks do
+most of the work, and a (unit) test implementation helps while testing.
+
+``` C#
+public interface IClock
+{
+    DateTime UtcNow();
+}
+```
+
+This is a common way of solving the issue. By using an abstract class instead,
+we can easily define the shared clock logic in the base class, as most of it
+is the same for all implementations:
+
+``` C#
+public abstract class Clock
+{
+	public static readonly Clock System = new SystemClock();
+
+	public abstract DateTime UtcNow();
+	
+	public DateOnly Today() => DateOnly.FromDateTie(UtcNow());
+	
+	public DateOnly Yesterday() => Today().AddDays(-1);
+	
+	public override ToString() => UtcNow().ToString("U");
+
+    private sealed class SystemClock : Clock
+	{
+	    public override DateTime UtcNow() => DateTime.UtcNow;
+	}
+}
+
+public sealed class TestClock
+	{
+		public TestClock(DateTime time) => Time = time;
+	    
+		private readonly DateTime Time;
+		
+		public override DateTime UtcNow() => Time;
+	}
+```
+
+## Singleton delegate
+The disadvantage of having an injectable date time provider is obviously that
+you have to pass it through. There is, however, an other way, and I prefer this
+one in almost all cases: A (testable) singleton delegate.
```

Comment:
In what case do you not prefer to use this? And does it have value to add it here as well?

Reply:
Well, if you create a library you might not want to add such a dependency to that package. I'm not sure how to write that down without making it a too big thing that distracts from the message.

Reply:
"such a dependency to that package" => you mean a dependency on a Inversion-of-control Container?
If a library depends on a ioc container, it is probably doing it wrong.
It might provide registration helper functions. But other than that, there is no need for a dependency to an ioc container.

Reply:
@Sjaaky What I ment was that having a dependency on an a mechanism (like `Qowaiv.Clock`) for a library is not preferable. Not to have a dependency on an IoC.